orangewidget/widget.py

```diff
@@ -611,16 +609,11 @@ def _():
                 b.clicked.connect(self.reset_settings)
                 buttonsLayout.addWidget(b)
             if hasattr(self, "set_visual_settings"):
-                assert self.initial_visual_settings is not None
-                dlg = VisualSettingsDialog(self)
-                dlg.initialize(self.initial_visual_settings)
-                dlg.setting_changed.connect(self.set_visual_settings)
-
                 icon = QIcon(gui.resource_filename("icons/visual-settings.svg"))
                 icon.addFile(gui.resource_filename("icons/visual-settings-hover.svg"),
                              mode=QIcon.Active)
                 b = SimpleButton(icon=icon, toolTip="Set visual settings")
-                b.clicked.connect(dlg.show_dlg)
+                b.clicked.connect(self.openVisualSettingsClicked)
```

Comment:
I support the change. I feared that enforcing a certain form of dialog is to rigid, but I forgot to write so.

In practice, the signal `openVisualSettingsClicked` will most often connect to a method that will do exactly the same as the lines you removed. Perhaps you could put them into some kind of utility function? I guess the function that contains the above lines could be a static function of `SettingsDialog`, something like

```python
@staticmethod
def attach(widget):
    assert widget.initial_visual_settings is not None
    dlg = SettingsDialog(widget, widget.initial_visual_settings)
    dlg.settings_changed.connect(widget.set_visual_settings)
    widget.openVisualSettings.clicked.connect(dlg.show_dlg)
```

I obviously haven't tested this. And I may have misread/misunderstood everything.

Reply:
Have you seen the __init__ in VisualSettingsDialog? It does what you proposed in the `attach`.